From 7a55344952601ddd08ade3269d4eaa4debcf322d Mon Sep 17 00:00:00 2001
From: Jordan Petridis <jpetridis@gnome.org>
Date: Fri, 12 Sep 2025 09:55:47 +0300
Subject: [PATCH] xdg-user-dirs: Backport systemd service patches

Backport of two patches so it will register as a service

https://gitlab.freedesktop.org/xdg/xdg-user-dirs/-/merge_requests/10

and

https://gitlab.freedesktop.org/xdg/xdg-user-dirs/-/merge_requests/12
---
 elements/components/xdg-user-dirs.bst |  5 +++
 patches/xdg-user-dirs/10.patch        | 57 +++++++++++++++++++++++++++
 patches/xdg-user-dirs/12.patch        | 51 ++++++++++++++++++++++++
 3 files changed, 113 insertions(+)
 create mode 100644 patches/xdg-user-dirs/10.patch
 create mode 100644 patches/xdg-user-dirs/12.patch

diff --git a/elements/components/xdg-user-dirs.bst b/elements/components/xdg-user-dirs.bst
index 767a8f817f..c468ac51c5 100644
--- a/elements/components/xdg-user-dirs.bst
+++ b/elements/components/xdg-user-dirs.bst
@@ -1,21 +1,26 @@
 kind: autotools
 
 build-depends:
 - public-stacks/buildsystem-autotools.bst
 - components/docbook-xsl.bst
 - components/libxslt.bst
+- components/systemd.bst
 
 depends:
 - public-stacks/runtime-minimal.bst
 
 variables:
   autogen: autoreconf -ivf -I %{datadir}/gettext/m4
   conf-link-args: ''
 
 sources:
 - kind: git_repo
   url: freedesktop:xdg/xdg-user-dirs.git
   track: '*.*'
   ref: 0.18-0-g52e71ff04416f2ca3adaf6d73193de4b8b83de46
 - kind: patch
   path: patches/xdg-user-dirs/gettext.patch
+- kind: patch
+  path: patches/xdg-user-dirs/10.patch
+- kind: patch
+  path: patches/xdg-user-dirs/12.patch
diff --git a/patches/xdg-user-dirs/10.patch b/patches/xdg-user-dirs/10.patch
new file mode 100644
index 0000000000..3d7bf224fe
--- /dev/null
+++ b/patches/xdg-user-dirs/10.patch
@@ -0,0 +1,57 @@
+From 2a63d3f0ffb76135790bb1168a3270a599904380 Mon Sep 17 00:00:00 2001
+From: David Redondo <kde@david-redondo.de>
+Date: Tue, 20 Sep 2022 10:37:51 +0200
+Subject: [PATCH] Add a systemd service to run xdg-user-dirs-update
+
+Recently GNOME and Plasma have been embracing systemd for handling
+their session startup. To enable autostart integration with this
+systemd includes systemd-xdg-autostart-generator which creates units
+from autostart desktop files. xdg-user-dirs installs such a desktop
+file (xdg-user-dirs.desktop). However because this file  contains
+X-GNOME-Autostart-Phase=Initialization systemd will skip it since
+it assumes it will be handles by GNOME itself. This a problem for
+Plasma and other desktop environments that do not handle this
+themselves, assuming that systemd takes care about autostart
+xdg-user-dirs-update will not be run.
+See https://github.com/systemd/systemd/issues/18791
+We can provide a systemd service file to make sure xdg-user-dirs-update
+is run during session startup. This ensures that the relevant
+directories of the user exist on login and during startup as
+specifying to be run before graphical-session-pre.target
+means that it runs sufficiently early in the startup process while
+using the generator would result in 'After=graphical-session.target'.
+---
+ xdg-user-dirs.desktop |  1 +
+ xdg-user-dirs.service | 11 +++++++++++
+ 2 files changed, 12 insertions(+)
+ create mode 100644 xdg-user-dirs.service
+
+diff --git a/xdg-user-dirs.desktop b/xdg-user-dirs.desktop
+index 6b969d4..b14a973 100644
+--- a/xdg-user-dirs.desktop
++++ b/xdg-user-dirs.desktop
+@@ -8,3 +8,4 @@ NoDisplay=true
+ 
+ X-GNOME-Autostart-Phase=Initialization
+ X-KDE-autostart-phase=1
++X-systemd-skip=true
+diff --git a/xdg-user-dirs.service b/xdg-user-dirs.service
+new file mode 100644
+index 0000000..6795045
+--- /dev/null
++++ b/xdg-user-dirs.service
+@@ -0,0 +1,11 @@
++[Unit]
++Description=User folders update
++Documentation=man:xdg-user-dirs-update(1)
++Before=graphical-session-pre.target
++
++[Service]
++Type=oneshot
++ExecStart=/usr/bin/xdg-user-dirs-update
++
++[Install]
++WantedBy=graphical-session-pre.target
+-- 
+GitLab
+
diff --git a/patches/xdg-user-dirs/12.patch b/patches/xdg-user-dirs/12.patch
new file mode 100644
index 0000000000..14f0e78c84
--- /dev/null
+++ b/patches/xdg-user-dirs/12.patch
@@ -0,0 +1,51 @@
+From 1512365de068446f8b5e14163601dd71e34e8bc1 Mon Sep 17 00:00:00 2001
+From: David Redondo <kde@david-redondo.de>
+Date: Thu, 9 Feb 2023 11:57:37 +0100
+Subject: [PATCH] Install systemd service file
+
+---
+ Makefile.am  | 4 +++-
+ configure.ac | 5 +++++
+ 2 files changed, 8 insertions(+), 1 deletion(-)
+
+diff --git a/Makefile.am b/Makefile.am
+index f649564..81d9976 100644
+--- a/Makefile.am
++++ b/Makefile.am
+@@ -14,7 +14,7 @@ INCLUDES =					\
+ 	$(NULL)
+ 
+ EXTRA_DIST= config.rpath translate.c autogen.sh \
+-	user-dirs.conf user-dirs.defaults xdg-user-dir xdg-user-dirs.desktop
++	user-dirs.conf user-dirs.defaults xdg-user-dir xdg-user-dirs.desktop xdg-user-dirs.service
+ 
+ xdgdir=$(sysconfdir)/xdg
+ xdg_DATA=user-dirs.conf user-dirs.defaults
+@@ -22,6 +22,8 @@ xdg_DATA=user-dirs.conf user-dirs.defaults
+ xdgautostartdir=$(xdgdir)/autostart
+ xdgautostart_DATA = xdg-user-dirs.desktop
+ 
++systemduserunit_DATA = xdg-user-dirs.service
++
+ libraries = $(LIBINTL)
+ 
+ bin_PROGRAMS =					\
+diff --git a/configure.ac b/configure.ac
+index 2cce082..dc389db 100644
+--- a/configure.ac
++++ b/configure.ac
+@@ -77,6 +77,11 @@ if test x$enable_documentation = xyes; then
+ fi
+ AM_CONDITIONAL(BUILD_DOCUMENTATION, test x$enable_documentation = xyes)
+ 
++PKG_CHECK_EXISTS([systemd],
++      [systemduserunitdir=$($PKG_CONFIG --variable=systemduserunitdir systemd)],
++      [systemduserunitdir='${prefix}/lib/systemd/user'])
++AC_SUBST(systemduserunitdir)
++
+ AC_OUTPUT([ po/Makefile.in
+ Makefile
+ man/Makefile
+-- 
+GitLab
+
-- 
2.50.1

